Feature/private abigen

zzy96 opened the pull request:
Enabling privateFor feature for sending private transactions via native go dapp

apratt3377 committed:
allow private transactions with abigen

apratt3377 committed:
add simulated abigen example

apratt3377 committed:
add comments and doc

apratt3377 committed:
fix comments

apratt3377 committed:
add comments

zzy96 committed:
private abigen merged

zzy96 committed:
Merge remote-tracking branch 'remotes/apratt/feature/privateAbigen' into feature/privateAbigen

Krish1979 commented on ethclient/ethclient.go:
```diff
@@ -490,6 +491,35 @@ func (ec *Client) SendTransaction(ctx context.Context, tx *types.Transaction) er
+// PreparePrivateTransaction sends the encoded raw transaction to Constellation,
+// returning the encoded commitment transaction.
+func (ec *Client) PreparePrivateTransaction(ctx context.Context, encodedTx hexutil.Bytes, privateFrom string, privateFor []string) (hexutil.Bytes, error) {
+	if len(privateFor) == 0 {
```
The privateFor object is NULL only for public transactions , but for transactions to itself it has a blank value. Checking length for zero will prevent a node sending private transactions to itself. So please change the check to != NULL rather..

vsmk98 replied:
changed to allow transaction to self node.

vsmk98 committed:
changes to allow private transaction to self node

vsmk98 committed:
changes to allow privateFor transaction to self node

zzy96 committed:
change error variable to err

zzy96 committed:
Merge branch 'master' into feature/privateAbigen

zzy96 committed:
cleaning & commenting...

zzy96 committed:
Merge branch 'master' into feature/privateAbigen

zzy96 committed:
auto format

zzy96 committed:
update comments

zzy96 committed:
Merge branch 'master' into feature/privateAbigen

zzy96 committed:
Merge branch 'master' into feature/privateAbigen

zzy96 committed:
Merge branch 'master' into feature/privateAbigen

zzy96 committed:
Merge branch 'master' into feature/privateAbigen

trung commented on core/types/transaction.go:
```diff
@@ -185,7 +186,11 @@ func (tx *Transaction) UnmarshalJSON(input []byte) error {
 
-func (tx *Transaction) Data() []byte       { return common.CopyBytes(tx.data.Payload) }
+func (tx *Transaction) Data() []byte { return common.CopyBytes(tx.data.Payload) }
+func (tx *Transaction) SetData(data []byte) {
```
IMO, transaction data is supposed to be immutable after transaction is created. Exposing `SetData`  breaks this design.

trung commented on ethclient/ethclient.go:
```diff
@@ -506,6 +507,35 @@ func (ec *Client) SendTransaction(ctx context.Context, tx *types.Transaction) er
+	}
+
+	tx.SetPrivate()
+	tx.SetData(data)
```
Use `types.NewTransaction` or `types.NewContractCreation` instead

zzy96 replied:
I have removed SetData and use types.NewTransaction/types.NewContractCreation to create private transaction

zzy96 committed:
Optimizing Code Changes

- remove PreparePrivateTransaction method in ContractTransactor interface
- remove SetData method in transaction.go
- use types.NewTransaction and types.NewContractCreation to create private transaction

zzy96 committed:
Merge branch 'master' into feature/privateAbigen

trung commented on accounts/abi/bind/base.go:
```diff
@@ -340,6 +356,30 @@ func (c *BoundContract) UnpackLog(out interface{}, event string, log types.Log)
+	if private.P == nil {
+		return nil, errors.New("transaction manager not set up")
+	}
+	data, err := private.P.Send(tx.Data(), privateFrom, privateFor)
```
this calls `/sendraw` which is only meant for Quorum node.
Native DApps can only call `/storeraw`

zzy96 committed:
Merge branch 'master' into feature/privateAbigen

zzy96 committed:
Flow changes

- use /storeraw to send private transaction payload
- use eth_sendRawPrivateTransaction API to send private transaction hash

zzy96 committed:
add SendPrivateTransaction to simulated backend

zzy96 committed:
parse JSON response from Tessera

zzy96 committed:
fix some errors

trung commented on accounts/abi/bind/base.go:
```diff
@@ -49,6 +53,10 @@ type TransactOpts struct {
 	Nonce  *big.Int       // Nonce to use for the transaction execution (nil = use pending state)
 	Signer SignerFn       // Method to use for signing the transaction (mandatory)
 
+	TransactionManagerURL string   // Tessera/Constellation URL to call /storeraw API.
```
`ethclient.Client` struct defines wrappers for Ethereum RPC API, it's the right place to define wrappers for Private Transaction Manager public API.

zzy96 replied:
This is quite tricky. It seems that we cannot create transaction in `ethclient.Client` if we want to use abigen go binding code to create transaction.

trung replied:
`ethclient.Client` implements `ContractBackend` interface and hence it is the implementation of `ContractTransactor` interface.

zzy96 replied:
Adding `TransactionManagerURL` as a field in `ethclient.Client` struct

trung commented on accounts/abi/bind/base.go:
```diff
@@ -340,6 +368,58 @@ func (c *BoundContract) UnpackLog(out interface{}, event string, log types.Log)
+	reqBodyJSON := &StoreRawArgs{
+		Payload: string(data),
+		From:    privateFrom}
+	reqBody, _ := json.Marshal(reqBodyJSON)
```
nitpick: return error if not nil

trung commented on accounts/abi/bind/base.go:
```diff
@@ -340,6 +368,58 @@ func (c *BoundContract) UnpackLog(out interface{}, event string, log types.Log)
+	defer resp.Body.Close()
+
+	respBodyJSON := StoreRawResp{}
+	json.Unmarshal(respBody, &respBodyJSON)
```
nitpick: return error if not nil

trung commented on accounts/abi/bind/base.go:
```diff
@@ -340,6 +368,58 @@ func (c *BoundContract) UnpackLog(out interface{}, event string, log types.Log)
+	if err != nil {
+		return nil, errors.New("Failed to call /storeRaw API on transaction manager.")
+	}
+	respBody, _ := ioutil.ReadAll(resp.Body)
```
nitpick: return error if not nil

trung commented on accounts/abi/bind/base.go:
```diff
@@ -340,6 +368,58 @@ func (c *BoundContract) UnpackLog(out interface{}, event string, log types.Log)
+		Payload: string(data),
+		From:    privateFrom}
+	reqBody, _ := json.Marshal(reqBodyJSON)
+	resp, err := http.Post(url+"/storeraw", "application/json", bytes.NewBuffer(reqBody))
```
Private Transaction Manager (Tessera) third-party API server supports [multiple ways of communications](http://docs.goquorum.com/en/latest/Privacy/Tessera/Configuration/Configuration%20Overview/#server). If `ethclient` only support HTTP, need to be explicit and handling check explicitly

zzy96 replied:
Make `ethclient` function name explicit as `StoreRawHTTP`

zzy96 committed:
add error handling

zzy96 committed:
fix missing base64 encoding/decoding. working now!

zzy96 committed:
code refactoring

zzy96 committed:
formatting & commenting

zzy96 committed:
change function name explicit with HTTP

zzy96 committed:
hard reset changes, make privateFrom optional in /storeraw

zzy96 committed:
rename StoreRawHTTP in ContractTransactor interface to PreparePrivateTransaction

zzy96 committed:
combine SendTransaction and SendPrivateTransaction into SendTransaction in ContractTransactor interface

zzy96 committed:
Merge branch 'master' into feature/privateAbigen

zzy96 committed:
fix unit test

zzy96 committed:
Merge branch 'feature/privateAbigen' of https://github.com/zzy96/quorum into feature/privateAbigen

zzy96 committed:
Merge branch 'master' into feature/privateAbigen

zzy96 committed:
Merge branch 'master' into feature/privateAbigen

trung commented on ethclient/ethclient.go:
```diff
@@ -34,7 +39,8 @@ import (
 type Client struct {
-	c *rpc.Client
+	c                         *rpc.Client
+	TransactionManagerHTTPURL string // Tessera/Constellation HTTP URL to call /storeraw API.
```
As discussed, please create a `PrivateTransactionManagerClient` to handle `/storeraw` API call

trung commented on ethclient/ethclient.go:
```diff
@@ -498,12 +504,55 @@ func (ec *Client) EstimateGas(ctx context.Context, msg ethereum.CallMsg) (uint64
+		return nil, errors.New("transaction manager url is nil")
+	}
+
+	reqBodyJSON := map[string]string{
```
Model `xxRequest` and `xxResponse` instead of generic map
